### DISCUSSION TITLE
Provide Custom Stream for Wasmedge SDK's stdin/stdout

### DISCUSSION THREAD

**Discussion by flosincapite** [WasmEdge/WasmEdge#3770] [Q&A] [answered]
Hello! I've been using the wasmedge CLI to inside a web service to run .wasm programs. Currently, the service is written in Node, so I run the wasmedge CLI in subprocesses. Crucially, the .wasm programs make heavy use of stdin/stdout, and I can't feasibly alter that behavior.

I'd like to switch to either the Rust or Go SDK in order to eliminate subprocesses. I'm able to run the .wasm binaries with both SDKs, but the main Rust/Go process's stdin/stdout are captured that way. I'd like to avoid having the top-level stdin/stdout get captured, not least because I need to run multiple copies of these binaries simultaneously.

Is there a way to hook WASI's std stream read/write behavior to some other construct (an in-memory stream, temporary file descriptor, etc.)?

**Comment by flosincapite** [comment]
The closest thing I am aware of is this feature request: https://github.com/WasmEdge/WasmEdge/issues/2936. If this feature were present in Rust or Go, it would work for my use case, but I am curious if there are more straightforward options.

**Reply by hydai** [reply, marked as answer]
#2936 should be the solution since we don't have any other way to redirect the studio fd.
